Align api with ffi

BitcoinZavior opened the pull request:
Major upgrade to align api with bdk-ffi

1 - Updated API to align with bdk-ffi [v0.11.0](https://github.com/bitcoindevkit/bdk-ffi/releases)
2- Updated make file to ensure build on different laptop configs and resolve bugs. 
3 - Update readme with build instructions in case someone wants to make changes to rust code and build.
(Points 3 and 4 are not relevant and not required for projects who intend to use bdk-flutter, this is relevant and useful for contributors to the project)
4 -  Comments added for auto doc generation, when the package gets published to pub.dev it will have complete api reference with mostly the same text and description as bdk.
5 - Updated test cases: https://github.com/LtbLightning/bdk-flutter/blob/align-api-with-ffi/test/bdk_flutter_test.dart
6- Updated Example App: https://github.com/LtbLightning/bdk-flutter/blob/align-api-with-ffi/example/lib/main.dart
7 - Kept MIT as the main license, this is to ensure pub. dev picks up the license information correctly. Kept Apache as an additional license.

Fixes #38
Fixes #37
Fixes #36
Fixes #35
Fixes #34
Fixes #33
Fixes #32
Fixes #27

thunderbiscuit commented on README.md:
```diff
@@ -161,21 +161,21 @@ A known `mnemonic` can be used when restoring a wallet, or a new `mnemonic` can
-                                                    blockchainConfig: BlockchainConfig.electrum
-						    (
-						        config: ElectrumConfig(
+final aliceWallet  =  await BdkFlutter().createWallet( mnemonic:"puppy interest whip tonight dad never sudden response push zone pig patch",
```
The spacing on most of the readme appears to be off on my local copy/IDE but not sure why.

<img width="1310" alt="spacing" src="https://user-images.githubusercontent.com/39974688/199566843-f81bb107-db4a-4565-a68f-ed3f45b6bea9.png">

thunderbiscuit commented on README.md:
```diff
@@ -89,12 +89,12 @@ The following methods can be used with this module. All methods can be called by
 ### generateMnemonic()
```
I think (or at least from what I can tell? let me know if I'm wrong) you've changed the `generateMnemonic()` method only take a wordcount argument, but forgot to fix it in the readme.

thunderbiscuit commented on lib/src/utils/helper.dart:
```diff
@@ -1,3 +1,5 @@
+import 'exceptions/wallet_exception.dart';
 
 const MIN_WORD_COUNT= 6;
 const MAX_WORD_COUNT = 24;
```
I think these `MIN_ENTROPY_BIT` and `MAX_ENTROPY_BIT` and `isBadEntropyBitCount()` have been removed because I can't find usage of them anywhere, so they could be removed here as well.

BitcoinZavior replied:
Yeah these are not required anymore. Will delete.

thunderbiscuit commented on rust/src/types.rs:
```diff
@@ -1,29 +1,99 @@
+    pub(crate) address: String,
+}
+#[derive(PartialEq)]
+pub enum DescriptorKeyType{
+    EXTENDED,
+    DERIVED,
+    DEFAULT
+}
```
What does this type do? I see it used in the DescriptorSecretKey type but I'm not sure I fully understand its purpose.

thunderbiscuit commented on rust/src/ffi.rs:
```diff
@@ -1,34 +1,24 @@
 use bdk::bitcoin::secp256k1::Secp256k1;
```
I recommend you run a formatter for your Rust code (maybe for the Dart code as well? I don't know if they are easy to use). You can just use `cargo fmt` and it will clean up a lot/most of the small nits you have (small spacing issues, etc.) over the Rust files.

BitcoinZavior replied:
Done

thunderbiscuit commented on ios/Classes/bindings.h:
```diff
@@ -41,38 +41,58 @@ typedef struct wire_BlockchainConfig {
+} wire_SqliteDbConfiguration;
 
 typedef struct DatabaseConfig_SQLITE {
-  struct wire_SqliteConfiguration *config;
```
Swift is not my specialty, but I'm wondering if you can tell me what the `wire` in the name at the beginning of all the arguments does?

BitcoinZavior replied:
That's autogenerated by flutter rust bridge

thunderbiscuit commented on rust/src/r_api.rs:
```diff
@@ -1,405 +1,529 @@
-use crate::ffi::{
```
Can you explain what is the architectural decision behind the `ffi.rs` file and the `r_api.rs` file? I sort of see them as both providing similar things and not sure what the distinction is between the two.

BitcoinZavior replied:
r_api.rs is the api that is exposed in Flutter. It uses ffi.rs (bdk-ffi). It does some structural wrapping to accommodate calls and code patterns that may not directly work from flutter if used with ffi.rs directly.
Also, adding some code to r_api(and root.dart) and removing from ffi.rs helps reduce the binary size. For example for txBuilder.

thunderbiscuit reviewed (commented):
It's a lot of code to review, but I added a few questions in places where I had questions.

It certainly looks like a good step forward.

BitcoinZavior commented on rust/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 name = "rust"
-version = "0.1.0"
+version = "0.1.6"
 edition = "2021"
```
That's automatically added when creating a new project, so I left it as is. Probably worth noting in case we find any issues related to this later.

BitcoinZavior commented on LICENSE-MIT:
```diff
@@ -1,16 +0,0 @@
-Permission is hereby granted, free of charge, to any person obtaining a copy of
```
Yep, there is a TODO for "Review and align license with BDK & LDK"
Also, pub.dev recommends we use an OSI-approved license from https://opensource.org/licenses.
So will have MIT license from https://opensource.org/licenses as the default license and add other license files in the repo from bdk

BitcoinZavior commented on example/lib/main.dart:
```diff
@@ -1,101 +1,122 @@
-    print("mnemonicWithEntropy: $mnemonicWithEntropy ");
-    print("mnemonicWithEntropy: $mnemonicWithEntropy2 ");
-    print("mnemonicWithEntropy: $mnemonicWithEntropy5 ");
+    AddressInfo d;
+    AddressIndex g;
+    SledDbConfiguration j;
+    SqliteDbConfiguration h;
+    TransactionDetails l;
+    BlockTime lg;
+    BlockchainConfig k;
+    TxOut i;
```
This was not meant to be committed, will delete in the next commit.

BitcoinZavior commented on rust/src/ffi.rs:
```diff
@@ -1,34 +1,24 @@
-    DerivableKey, DescriptorPublicKey as BdkDescriptorPublicKey,
-    DescriptorSecretKey as BdkDescriptorSecretKey, ExtendedKey, GeneratableKey, GeneratedKey,
-};
+use bdk::keys::{DerivableKey, DescriptorPublicKey as BdkDescriptorPublicKey, DescriptorSecretKey as BdkDescriptorSecretKey, ExtendedKey, GeneratableKey, GeneratedKey};
```
Yes, possibly this may not be required. Worth trying it out. In the past I have used BDK directly.
In the current version I wanted to reuse most of bdk-ffi/lib.rs as is

BullishNode commented:
Had a look at the readme. The api is much more powerful now. Many requested features available. Thx so much for your work 🧡

notmandatory commented:
I took a quick look through `ffi.rs` and `r_api.rs` and this looks like a big step in the right direction. Your use of `Opaque` make this feel much more like the `bdk-ffi` apis. I do see you're using `.panic()` for error handling, eventually you should probably handle these cases with returned errors, but I understand if you want to get this out so users can start playing with it. One small nit is your naming of `WalletInstance` and `BlockchainInstance` I'd call them simply `Wallet` and `Blockchain`, but certainly either way works.

BitcoinZavior commented:
> One small nit is your naming of `WalletInstance` and `BlockchainInstance` I'd call them simply `Wallet` and `Blockchain`, but certainly either way works.

Yes, I did change these names a couple of times as I couldn't decide on the most appropriate ones. I think your suggestion is a good one, I have just addressed it and updated all `*Instance` to `*`

> I do see you're using .panic() for error handling, eventually you should probably handle these cases with returned errors, but I understand if you want to get this out so users can start playing with it.

rust_bridge converts it all into ffiexception, we catch and throw a custom exception, but its somewhat basic, so doesn't cover all the cases. Perhaps going through bdk docs and handling exceptions more efficiently will be a good exercise. Opening an issue for this to come up with a better approach, https://github.com/LtbLightning/bdk-flutter/issues/40.
